dask/dataframe/groupby.py

```diff
@@ -1866,6 +1891,55 @@ def unique(self, split_every=None, split_out=1):
         )
 
 
+def _nlargest_chunk(df, *index, **kwargs):
+    n = kwargs.pop("n")
+    keep = kwargs.pop("keep")
+    name = kwargs.pop("name")
+    indices = list(index)
+    series_grouped = df.groupby(indices)[name].nlargest(n, keep)
+    if len(series_grouped) == len(df):
+        df_grouped = series_grouped.to_frame()
+        num_indices = len(indices)
+
+        all_missing_columns = []
+        for i in range(num_indices):
+            if indices[i] == name:
+                pass
+            else:
+                temp = df[indices].iloc[:, i]
+                all_missing_columns.append(temp)
+
+        all_missing_columns.append(df.index)
+
+        return df_grouped.set_index(all_missing_columns)[name]
+    else:
+        if name in indices:
+            return series_grouped.reset_index(name, drop=True)
+        return series_grouped
+
+
+def _nlargest_aggregate(df, **kwargs):
+    n = kwargs.pop("n")
+    keep = kwargs.pop("keep")
+    name = kwargs.pop("name")
+    index = kwargs.pop("index")
+    num_records = kwargs.pop("len")
+    indices = list(index)
+    df = df.reset_index()
+    level_num = list(df.columns).index(name) - 1
```

Comment:
`name` is always in `df.columns`? Or can it be referring to the level of an index? Pandas (and I think Dask) supports either.

Reply:
Yes, the `name` should be present in `df.columns` always.
In the operation, `dask_df.groupby(["location", "category"])["percent"].nlargest(2)`, the name is `percent`. It represents the column for which we want these `nlargest` values.

Since this is always passed before calling `.nlargest(K)`, thus the `percent` column will be always present in the result (irrespective of if it's a `MultiIndex` or a `RangeIndex`). Now, when operations such as `to_frame()` OR perhaps `.reset_index()` are done over it, we get the column `percent` back, which was hidden before (in a `MultiIndex` or a `RangeIndex`)

Since the operation `.reset_index()` is applied before calculating this `level_num` value, we are bound to get a valid value for this variable.

This is expected because the `nlargest` function that we implemented here is for `groupby.SeriesGroupBy.nlargest` object (and not for `DataFrame.nlargest` or `Series.nlargest` -- they are already implemented in Dask) which can be only called after we access a column first (in this case the `percent` column).